### PR TITLE
feat(engine): cross-session prompt caching to reduce per-task token consumption

### PR DESCRIPTION
**Pull request by magreenberg1**
## Problem

A single multi-step Fusion task can consume 500K+ tokens, exhausting Claude Pro session limits in 30–60 minutes. The majority of this cost is **overhead**, not useful work:

- **~18K tokens of skill references** loaded for every session — including reviewer and heartbeat sessions that don't need executor-focused skills (71KB total across 12 reference files)
- **System prompt prefix varies between sessions** of the same role, preventing Anthropic's prompt cache (90% read discount) and OpenAI's automatic prefix cache (50% discount) from activating
- **Tool ordering is non-deterministic**, breaking cache prefix matching even when the tools themselves are identical

For a typical 5-step task at review level 2 (13 sessions), overhead alone accounts for ~350K–400K tokens before any implementation work begins.

## Solution

Three complementary changes that reduce per-task token overhead by an estimated 50–60%:

### 1. Prompt layering (`SystemPromptLayers`)

Factors the monolithic system prompt string into structured layers:

```
┌─────────────────────────────┐
│  stable (cacheable prefix)  │  ← Base role prompt + memory instructions
│  ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─  │     Byte-identical across all sessions of the same role
│  dynamic (per-session)      │  ← Agent instructions, plugins, performance feedback
└─────────────────────────────┘
```

The stable layer is passed to pi-coding-agent's `systemPromptOverride`, and the dynamic layer goes through `appendSystemPromptOverride`. This separates the cacheable prefix from per-session content at the API boundary.

Applied to all four session types: **executor**, **reviewer**, **triage**, and **heartbeat**.

### 2. Skill scoping by session purpose

`getSkillPurposeFilter(sessionPurpose)` returns a predicate that decides which skills to load:

| Session purpose | Skills loaded |
|----------------|---------------|
| `executor`, `triage` | All skills (pass-through) |
| `reviewer` | `code-review`, `security-review`, `review` only |
| `heartbeat` | `monitoring`, `heartbeat` only |
| Unknown | All skills (safe fallback) |

This prevents reviewer sessions from loading ~18K tokens of executor-focused skill references they never use.

### 3. Deterministic tool ordering

Tools are sorted alphabetically by name after the full tool chain is assembled in `createFnAgent`. This ensures the tool schema portion of the API request is byte-identical across sessions, which is a prerequisite for prompt cache prefix matching.

### Bonus: Cache hit ratio metric

`computeCacheHitRatio(inputTokens, cachedTokens)` computes the fraction of input tokens served from cache. Exported from `session-token-usage.ts` for dashboard/observability use.

## Estimated token savings (5-step task, review level 2, 13 sessions)

| Change | Mechanism | Tokens saved |
|--------|-----------|-------------|
| Skill scoping | 10 reviewer sessions × ~18K tokens each | ~180K |
| Prompt caching | Shared prefix at 90% discount after first session | ~80–120K |
| Deterministic tools | Enables cache prefix matching for tool schemas | Prerequisite |
| **Total** | | **~260–300K (~50–60% of overhead)** |

### New files
- `packages/engine/src/prompt-layers.ts` — `SystemPromptLayers` type, `buildPromptLayers()`, `collapsePromptLayers()`

### Modified files
- `packages/engine/src/agent-runtime.ts` — `systemPromptLayers?: SystemPromptLayers` field on `AgentRuntimeOptions`
- `packages/engine/src/pi.ts` — `DefaultResourceLoader` config uses layers; tools sorted alphabetically
- `packages/engine/src/reviewer.ts` — Layered prompt assembly + `systemPromptLayers` passed to session
- `packages/engine/src/executor.ts` — Same pattern (2 call sites: initial + retry)
- `packages/engine/src/triage.ts` — Same pattern (2 call sites: initial + fallback)
- `packages/engine/src/agent-heartbeat.ts` — Same pattern
- `packages/engine/src/skill-resolver.ts` — `getSkillPurposeFilter()` export
- `packages/engine/src/session-token-usage.ts` — `computeCacheHitRatio()` export

## Backward compatibility

All changes are backward compatible:

- `systemPromptLayers` is **optional** on `AgentRuntimeOptions`. Runtimes that don't understand it use the existing `systemPrompt` string — behavior is identical to before when layers are absent.
- `collapsePromptLayers(buildPromptLayers({...}))` produces **byte-identical output** to the old `buildSystemPromptWithInstructions()` + manual concatenation. This is verified by 9 backward compatibility tests covering all four subsystem patterns.
- Skill scoping returns a **pass-all filter** for unknown session purposes, so new session types degrade safely.
- Tool sorting is stable and doesn't affect tool behavior — only ordering in the API payload.

## Test plan

61 tests across 12 test files, all passing:

| Test file | Tests | What it verifies |
|-----------|-------|-----------------|
| `prompt-layers.test.ts` | 11 | Core `buildPromptLayers`/`collapsePromptLayers` logic |
| `prompt-layers-backward-compat.test.ts` | 9 | Collapsed layers match legacy output for all 4 subsystems |
| `pi-layers-wiring.test.ts` | 8 | `createFnAgent` passes layers to `DefaultResourceLoader`; tools sorted in `createAgentSession` |
| `prompt-cache-integration.test.ts` | 4 | Byte-identical stable prefix across 10 simulated reviewer sessions |
| `skill-resolver-scoping.test.ts` | 5 | `getSkillPurposeFilter` per-role filtering |
| `skill-scoping-integration.test.ts` | 9 | Realistic 10-skill filtering; disjoint sets; executor superset |
| `token-usage-cache-ratio.test.ts` | 4 | `computeCacheHitRatio` edge cases |
| `reviewer-prompt-layers.test.ts` | 3 | Reviewer stable layer isolation |
| `pi-prompt-layers.test.ts` | 3 | Config branching logic |
| `agent-runtime-layers.test.ts` | 2 | Type accepts `systemPromptLayers` |
| `tool-ordering.test.ts` | 2 | Sort correctness and stability |
| `session-helpers-layers.test.ts` | 1 | `ResolvedSessionOptions` forwards layers |

```bash
cd packages/engine && pnpm exec vitest run src/__tests__/prompt-layers*.test.ts src/__tests__/pi-*.test.ts src/__tests__/reviewer-prompt-layers.test.ts src/__tests__/tool-ordering.test.ts src/__tests__/skill-*.test.ts src/__tests__/token-usage-cache-ratio.test.ts src/__tests__/agent-runtime-layers.test.ts src/__tests__/session-helpers-layers.test.ts
```

## How to verify cache effectiveness

1. Run a 3+ step task at review level 2
2. Check `tokenUsage` via `fn task show <id> --json | jq '.tokenUsage'`
3. Compare `cachedTokens` vs `inputTokens` — higher ratio = caching working
4. Check [Anthropic console usage](https://console.anthropic.com/settings/usage) for `cache_read` vs `input` token breakdown

## Summary by CodeRabbit

* **New Features**
  * Structured prompt-layer support (stable + dynamic) across agent flows for better cross-session caching
  * Token cache-hit ratio utility
  * Skill filtering by session purpose

* **Improvements**
  * Executor/reviewer/triage/heartbeat flows now produce and pass layered prompts while retaining legacy collapsed strings
  * Deterministic, alphabetized tool ordering; heartbeat completion tool kept last

* **Tests**
  * Extensive new tests for prompt layers, backward compatibility, cache behavior, and tool-ordering determinism

[Review Change Stack](https://app.coderabbit.ai/change-stack/Runfusion/Fusion/pull/63)

### CODE REVIEW

**Comment by gsxdsm** [comment]
This is amazing thank you

**Comment by gsxdsm** [comment]
Can you take a look at the typecheck and test failures? Might need to just sync latest from main. Will get this merged in asap

**Comment by magreenberg1** [comment]
@gsxdsm - I'll take a look tomorrow! Though my wife is 41 weeks pregnant, so if I disappear for a while that would be why!

**Comment by gsxdsm** [comment]
Congratulations!!!! No rush

File: packages/engine/src/skill-resolver.ts [hunk 1]
```diff
@@ -484,3 +484,46 @@ export function createSkillsOverrideFromSelection(
...
+    case "reviewer":
+      return (name) => REVIEWER_SKILL_ALLOWLIST.has(name);
+    case "heartbeat":
+      return (name) => HEARTBEAT_SKILL_ALLOWLIST.has(name);
+    case "executor":
+    case "triage":
+    default:
+      return () => true;
+  }
+}
```
[the remark this answers is not in the document]
Comment (magreenberg1) [reply packages/engine/src/skill-resolver.ts]: Fixed in d8a26ef2. You're right — `getSkillPurposeFilter` was dead code never wired into the actual skill-loading path (`buildSessionSkillContext` → `ROLE_FALLBACK_SKILLS`). Removed the function, both allowlists, and the two test files. The PR description's 180K skill-scoping savings claim was incorrect and has been corrected.

File: packages/engine/src/reviewer.ts [hunk 2]
```diff
@@ -408,23 +408,27 @@ export async function reviewStep(
...
-    : reviewerSystemPrompt;
+  if (reviewerPluginContributions) {
+    reviewerLog.log(`applied plugin prompt contributions for reviewer surface`);
+  }
+
+  const layers = buildPromptLayers({
+    basePrompt: reviewerBasePrompt + memorySection,
+    agentInstructions: reviewerInstructions,
+    pluginContributions: reviewerPluginContributions,
+  });
```
[the remark this answers is not in the document]
Comment (magreenberg1) [reply packages/engine/src/reviewer.ts:433]: Fixed in d8a26ef2. Moved `memorySection` out of `basePrompt` (stable) and into the dedicated `memorySection` parameter of `buildPromptLayers` (dynamic). The stable prefix is now just the base reviewer prompt, so it stays byte-identical even if agent memory is updated mid-task. Updated all related tests.

File: packages/engine/src/session-token-usage.ts [hunk 3]
```diff
@@ -90,3 +90,21 @@ export async function accumulateSessionTokenUsage(
...
+ * @returns Cache hit ratio in [0, 1], or 0 if no tokens used
+ */
+export function computeCacheHitRatio(
+  inputTokens: number,
+  cachedTokens: number,
+): number {
+  const total = inputTokens + cachedTokens;
+  if (total === 0) return 0;
+  return cachedTokens / total;
+}
```
[the remark this answers is not in the document]
Comment (magreenberg1) [reply packages/engine/src/session-token-usage.ts:110]: Fixed in d8a26ef2. Updated the JSDoc to clarify that `inputTokens` should be `tokens.input` only (NOT including `tokens.cacheWrite`), and that the returned ratio matches Anthropic console's `cache_read / input` metric.

File: packages/engine/src/pi.ts [hunk 4]
```diff
@@ -1748,6 +1757,10 @@ export async function createFnAgent(options: AgentOptions): Promise<AgentResult>
       boundaryContext.worktreePath,
       boundaryContext.worktreeProjectRoot,
     );
+    // Sort tools alphabetically by name for deterministic ordering.
+    // Prompt caching requires the tool list to be byte-identical across
+    // sessions — reordering breaks cache prefix matching.
+    customToolList.sort((a, b) => a.name.localeCompare(b.name));
```
[the remark this answers is not in the document]
Comment (magreenberg1) [reply packages/engine/src/pi.ts:1765]: Fixed in 082a79fb. The sort now preserves `fn_heartbeat_done` at the end by splicing it out after sorting and pushing it back. Verified with all 115 heartbeat executor tests passing, including the explicit `expect(toolNames[toolNames.length - 1]).toBe('fn_heartbeat_done')` assertion.

File: packages/engine/src/reviewer.ts [hunk 5]
```diff
@@ -408,23 +408,29 @@ export async function reviewStep(
...
+  const layers = buildPromptLayers({
+    basePrompt: reviewerBasePrompt,
+    agentInstructions: reviewerInstructions,
+    memorySection,
+    pluginContributions: reviewerPluginContributions,
+  });
+
+  // Collapsed string for backward compatibility with runtimes that don't
+  // support layers (plugin runtimes, older pi versions).
+  const reviewerSystemPromptFinal = collapsePromptLayers(layers);
```
[the remark this answers is not in the document]
Comment (magreenberg1) [reply packages/engine/src/reviewer.ts:437]: Fixed in a8e01237. Reordered `buildPromptLayers` to emit `memorySection` before `agentInstructions` in the dynamic layer, preserving the legacy ordering where memory preceded instructions. Added an explicit ordering assertion in the backward compat test (`memoryIdx < instructionsIdx < pluginsIdx`).

File: packages/engine/src/reviewer.ts [hunk 6]
```diff
@@ -408,23 +408,29 @@ export async function reviewStep(
   const memorySection = options.rootDir && options.settings?.memoryEnabled !== false
     ? "\n" + buildReviewerMemoryInstructions(options.rootDir, options.settings)
```
[the remark this answers is not in the document]
Comment (magreenberg1) [reply packages/engine/src/reviewer.ts]: Fixed in ee442db5. Removed the leading `"\n"` from the `memorySection` assignment in reviewer.ts. The `"\n"` was a legacy separator needed when memory was concatenated directly onto `basePrompt` — now that memory goes through `buildPromptLayers`, the layer joining handles separation with `"\n\n"`. Also rebased onto latest main (v0.26.0).

**Comment by gsxdsm** [comment]
The test failures are okay to ignore. I’m fixing those upstream.

**Comment by gsxdsm** [comment]
If you can address the compute cache hit ratio comment I’ll merge this in thanks!

File: packages/engine/src/session-token-usage.ts [hunk 7]
```diff
@@ -90,3 +90,23 @@ export async function accumulateSessionTokenUsage(
...
+ *   or 0 if no tokens used
+ */
+export function computeCacheHitRatio(
+  inputTokens: number,
+  cachedTokens: number,
+): number {
+  const total = inputTokens + cachedTokens;
+  if (total === 0) return 0;
+  return cachedTokens / total;
+}
```
[the remark this answers is not in the document]
Comment (magreenberg1) [reply packages/engine/src/session-token-usage.ts:110]: Fixed in 1880db08. Aligned the JSDoc with the stored `task.tokenUsage` format — `inputTokens` includes cache-write tokens (matching `accumulateSessionTokenUsage` line 55). Added a note that this differs from the Anthropic console metric which excludes cache-write from the denominator.

**Review by gsxdsm — APPROVED** [review]

**Comment by gsxdsm** [comment]
Thank you!!

**Comment by magreenberg1** [comment]
@gsxdsm - Thank you! Big fan of what you're doing - Always happy to help where possible!
